eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/ZmqWatcher.scala

```diff
@@ -44,157 +42,335 @@ import scala.util.Try
  *  - receives bitcoin events (new blocks and new txs) directly from the bitcoin network
  *  - also uses bitcoin-core rpc api, most notably for tx confirmation count and block count (because reorgs)
  */
-class ZmqWatcher(chainHash: ByteVector32, blockCount: AtomicLong, client: ExtendedBitcoinClient)(implicit ec: ExecutionContext = ExecutionContext.global) extends Actor with ActorLogging {
+object ZmqWatcher {
 
-  import ZmqWatcher._
+  // @formatter:off
+  sealed trait Command
+  sealed trait Watch extends Command
+  sealed trait WatchTriggered
+  private case class TriggerEvent[T <: WatchTriggered](replyTo: ActorRef[T], watch: Watch, event: T) extends Command
```

Comment:
I'll try to get rid of this generic type and see if we can avoid the explicit extraction of `replyTo` via exhaustive pattern matching.

Reply:
I can't find a satisfying solution here...
Here are the two approaches I can see:

1. Updates watches to have a `replyTo: ActorRef[WatchTriggered]` instead of a narrower type: if we do that the code inside `ZmqWatcher` becomes simpler, but external actors will need a downcast to ensure they receive the right concrete `WatchTriggered`
2. Watches have concrete `replyTo` types (the current state of the PR): since `ActorRef` is contravariant, we can't safely upcast those to a common `ActorRef[WatchTriggered]` so we need to enumerate each case

I can't find a best-of-both-world approach and didn't find anything in the akka best practices.